drivers/nrf24l01p/include/nrf24l01p_states.h

```diff
@@ -0,0 +1,182 @@
+/*
+ * Copyright (C) 2019 Otto-von-Guericke-Universität Magdeburg
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+/**
+ * @ingroup drivers_nrf24l01p
+ * @{
+ *
+ * @file
+ * @brief   Operation states of NRF24L01P devices
+ *
+ * @author  Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author  Peter Kietzmann <peter.kietzmann@haw-hamburg.de>
+ * @author  Fabian Hüßler <fabian.huessler@ovgu.de>
+ */
+#ifndef NRF24L01P_STATES_H
+#define NRF24L01P_STATES_H
+
+#include "nrf24l01p.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief   Transition delay to POWER_ON state
+ */
+#define NRF24L01P_DELAY_US_POWER_ON_RESET   (100000)
+
+/**
+ * @brief   Wakeup delay
+ */
+#define NRF24L01P_DELAY_US_START_UP         (1500)
+
+/**
+ * @brief   Transition delay to RX mode
+ */
+#define NRF24L01P_DELAY_US_RX_SETTLING      (130)
+
+/**
+ * @brief   Transition delay to TX mode
+ */
+#define NRF24L01P_DELAY_US_TX_SETTLING      (130)
+
+/**
+ * @brief   CE gpio pin pull-low time to trigger the transmission of a frame
+ */
+#define NRF24L01P_DELAY_US_CE_HIGH_PULSE    (10)
+
+/**
+ * @brief   Translate an NRf24L01P state to a netif state
+ *
+ * @param[in] state     NRF24L01P state
+ *
+ * @return              netopt state
+ * @retval -ENOTSUP     Unsupported state
+ */
+netopt_state_t nrf24l01p_state_to_netif(nrf24l01p_state_t state);
+
+/**
+ * @brief   Translate a netif state to an NRf24L01P state
+ *
+ * @param[in] state     netopt state
+ *
+ * @return              NRF24L01P state
+ * @retval -ENOTSUP     Unsupported state
+ */
+nrf24l01p_state_t nrf24l01p_state_from_netif(netopt_state_t state);
+
+#if !defined (NDEBUG) || defined(DOXYGEN)
+/**
+ * @name NRF24L01P Transitions
+ *
+ * Decode possible state transitions
+ * @{
+ */
+/**
+ * @brief Transitions that can be made from state UNDEFINED
+ */
+#define NRF24L01P_TRANSITIONS_FROM_UNDEFINED    (NRF24L01P_STATE_POWER_DOWN)
+
+/**
+ * @brief Transitions that can be made from state POWER_DOWN
+ */
+#define NRF24L01P_TRANSITIONS_FROM_POWER_DOWN   (NRF24L01P_STATE_POWER_DOWN  | \
+                                                 NRF24L01P_STATE_STANDBY_1)
+/**
+ * @brief Transitions that can be made from state STANDBY_1
+ */
+#define NRF24L01P_TRANSITIONS_FROM_STANDBY_1    (NRF24L01P_STATE_STANDBY_1   | \
+                                                 NRF24L01P_STATE_POWER_DOWN  | \
+                                                 NRF24L01P_STATE_STANDBY_2   | \
+                                                 NRF24L01P_STATE_RX_MODE     | \
+                                                 NRF24L01P_STATE_TX_MODE)
+/**
+ * @brief Transitions that can be made from state STANDBY_2
+ */
+#define NRF24L01P_TRANSITIONS_FROM_STANDBY_2    (NRF24L01P_STATE_STANDBY_2   | \
```

Comment:
Without knowing the state machine of the device: Shouldn't it be possible to transition between the two standby states? Otherwise it would not be possible from state 2 to wake up for RX without a full power cycle.

Reply:
You con go from `STANDBY_1` to `STANDBY_2`, but not directly from `STANDBY_2` to `STANDBY_1`.
From `STANDBY_2` you can go to `POWER_DOWN`, then `STANDBY_1`.
Or from `STANDBY_2` you can go to `TX_MODE`, then `STANDBY_1`.

The `STANDBY_2` is some kind of weird state and the datasheet distinguish between "recommended" and "possible" states, where `STANDBY_2` is the only state, not being recommended.

You go to `STANDBY_2` if you wanted to go to `TX_MODE` but you have actually not put anything to be transmitted in the FIFO. And in `STANDBY_2`, you basically wait until something is written to the TX FIFO.

Since it is not recommended, I don´t use it in the driver. But for completeness I wrote the `#define`. I think It could save some power if a device is only supposed to act as a transmitter, thus will never receive anything. I would call `STANDBY_2` something like a "pseudo TX state" because you just go there when you are configured as a transmitting device and have nothing in your FIFO